zrml/court/src/lib.rs

```diff
@@ -27,221 +27,1131 @@ extern crate alloc;
+                    ensure!(!reassigned, Error::<T>::JurorsAlreadyReassigned);
+                    ensure!(punished, Error::<T>::PunishTardyJurorsFirst);
+                    winner
+                }
+                _ => return Err(Error::<T>::CourtNotClosed.into()),
+            };
+
+            let draws = Draws::<T>::get(market_id);
+
+            let mut valid_winners_and_losers = Vec::with_capacity(draws.len());
+
+            for draw in draws {
+                if let Vote::Revealed { secret: _, outcome, salt: _ } = draw.vote {
+                    valid_winners_and_losers.push((draw.juror, outcome, draw.slashable));
+                }
+            }
+
+            Self::slash_losers_to_award_winners(
+                &market_id,
+                valid_winners_and_losers.as_slice(),
+                &winner,
+            );
+
+            court.status = CourtStatus::Closed { winner, punished: true, reassigned: true };
+            <Courts<T>>::insert(market_id, court);
+            <Draws<T>>::remove(market_id);
+
+            Self::deposit_event(Event::JurorStakesReassigned { market_id });
+
+            Ok(())
+        }
+    }
 
     impl<T> Pallet<T>
     where
         T: Config,
     {
-        // Returns an unique random subset of `jurors` with length `len`.
-        //
-        // If `len` is greater than the length of `jurors`, then `len` will be capped.
-        pub(crate) fn random_jurors<'a, 'b, R>(
-            jurors: &'a [(T::AccountId, Juror)],
-            len: usize,
+        pub(crate) fn choose_multiple_weighted<R: RngCore>(
+            jurors: &[JurorPoolItemOf<T>],
+            number: usize,
             rng: &mut R,
-        ) -> ArrayVec<&'b (T::AccountId, Juror), MAX_RANDOM_JURORS>
-        where
-            'a: 'b,
-            R: RngCore,
-        {
-            let actual_len = jurors.len().min(len);
-            jurors.choose_multiple(rng, actual_len).collect()
+        ) -> Result<Vec<DrawOf<T>>, DispatchError> {
+            let total_weight = jurors
+                .iter()
+                .map(|pool_item| {
+                    pool_item.stake.saturating_sub(pool_item.slashed).saturated_into::<u128>()
+                })
+                .sum::<u128>();
+
+            let mut random_set = BTreeSet::new();
+            let mut insert_unused_random_number = || -> DispatchResult {
+                let mut count = 0u8;
+                // this loop is to make sure we don't insert the same random number twice
+                while !random_set.insert(rng.gen_range(0u128..=total_weight)) {
+                    count = count.saturating_add(1u8);
+                    if count >= 3u8 {
+                        return Err(Error::<T>::RandNumGenFailed.into());
+                    }
+                }
+                Ok(())
+            };
+
+            for _ in 0..number {
+                insert_unused_random_number()?;
+            }
```

Comment:
`debug_assert!(random_set.len() == number)`

Reply:
It's just a reminder for myself to ensure that we have exactly the required random numbers

Reply:
09860bdd9613336504393f0dd4372dff7f18e2ad